File: Python/Practice Set 2/Dynamic Programming/DP51 - Burst Balloons.py
# Problem link - https://www.naukri.com/code360/problems/mining-diamonds_4244494?source=youtube&campaign=striver_dp_videos
# Solution - https://www.youtube.com/watch?v=Yz4LlDSlkns&list=PLgUwDviBIf0qUlt5H_kiKYaNSqJ81PMMY&index=52


def recursive():
    """
        Time complexity is exponential and space complexity is polynomial.
    """
    def solve(arr, i, j):
        if i > j:
            return 0
        max_coins = 0
        for index in range(i, j + 1):
            coins = arr[i - 1] * arr[index] * arr[j + 1] + solve(arr, i, index - 1) + solve(arr, index + 1, j)
            max_coins = max(max_coins, coins)
        return max_coins

    def burst(arr):
        n = len(arr)
        balloons = [1, ] + arr + [1, ]
        return solve(balloons, 1, n)

    print(burst([3, 1, 5, 8]))
    print(burst([7, 1, 8]))
    print(burst([9, 1]))
    print(burst([1, 2, 3, 4, 5]))
    print(burst([1, 5, 2, 8]))
    print(burst([5, 10]))
    print(burst([1, 5]))


def memoized():
    """
        Time complexity is O(n^3) and space complexity is O(n^2 + n).
    """
    def solve(arr, i, j, dp):
        if i > j:
            return 0
        if dp[i][j] is not None:
            return dp[i][j]
        max_coins = 0
        for index in range(i, j + 1):
            coins = arr[i - 1] * arr[index] * arr[j + 1] + solve(arr, i, index - 1, dp) + solve(arr, index + 1, j, dp)
            max_coins = max(max_coins, coins)
        dp[i][j] = max_coins
        return dp[i][j]

    def burst(arr):
        n = len(arr)
        balloons = [1, ] + arr + [1, ]
        dp = {i: {j: None for j in range(1, n + 1)} for i in range(1, n + 1)}
        return solve(balloons, 1, n, dp)

    print(burst([3, 1, 5, 8]))
    print(burst([7, 1, 8]))
    print(burst([9, 1]))
    print(burst([1, 2, 3, 4, 5]))
    print(burst([1, 5, 2, 8]))
    print(burst([5, 10]))
    print(burst([1, 5]))


def tabulation():
    """
        Time complexity is O(n^3) and space complexity is O(n^2).
    """
    def burst(arr):
        n = len(arr)
        arr = [1, ] + arr + [1, ]
        dp = {i: {j: 0 for j in range(n + 2)} for i in range(n + 2)}
        for i in range(n, 0, -1):
            for j in range(1, n + 1):
                if i > j:
                    continue
                max_coins = 0
                for index in range(i, j + 1):
                    coins = (arr[i - 1] * arr[index] * arr[j + 1]) + dp[i][index - 1] + dp[index + 1][j]
                    max_coins = max(max_coins, coins)
                dp[i][j] = max_coins
        return dp[1][n]

    print(burst([3, 1, 5, 8]))
    print(burst([7, 1, 8]))
    print(burst([9, 1]))
    print(burst([1, 2, 3, 4, 5]))
    print(burst([1, 5, 2, 8]))
    print(burst([5, 10]))
    print(burst([1, 5]))


recursive()
print()
memoized()
print()
tabulation()

File: Java/DynamicProgramming/DP51 - Burst Balloons.java
package DynamicProgramming;


import java.util.*;


class RecursiveDP51 {
    /**
     * Time complexity is exponential and space complexity is O(n).
     */
    private static Integer solve(List<Integer> arr, Integer i, Integer j) {
        if (i > j) {
            return 0;
        }
        Integer maxCoins = Integer.MIN_VALUE;
        for (int index = i; index <= j; index += 1) {
            Integer coins = arr.get(i - 1) * arr.get(index) * arr.get(j + 1) + solve(arr, i, index - 1) + solve(arr, index + 1, j);
            maxCoins = Math.max(coins, maxCoins);
        }
        return maxCoins;
    }

    public static Integer getMaximumCoins(List<Integer> balloons) {
        int n = balloons.size();
        List<Integer> arr = new ArrayList<>();
        arr.add(1);
        arr.addAll(balloons);
        arr.add(1);
        return solve(arr, 1, n);
    }

    public static void main(String[] args) {
        System.out.println(getMaximumCoins(Arrays.asList(3, 1, 5, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(7, 1, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(9, 1)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 2, 3, 4, 5)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5, 2, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(5, 10)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5)));
    }
}

class MemoizedDP51 {
    /**
     * Time complexity is O(n^3) and space complexity is O(n + n^2).
     */
    private static Integer solve(List<Integer> arr, Integer i, Integer j, Map<Integer, Map<Integer, Integer>> dp) {
        if (i > j) {
            return 0;
        }
        if (dp.get(i).get(j) != null) {
            return dp.get(i).get(j);
        }
        Integer maxCoins = Integer.MIN_VALUE;
        for (int index = i; index <= j; index += 1) {
            Integer coins = arr.get(i - 1) * arr.get(index) * arr.get(j + 1) + solve(arr, i, index - 1, dp) + solve(arr, index + 1, j, dp);
            maxCoins = Math.max(coins, maxCoins);
        }
        Map<Integer, Integer> mp = dp.get(i);
        mp.put(j, maxCoins);
        dp.put(i, mp);
        return dp.get(i).get(j);
    }

    public static Integer getMaximumCoins(List<Integer> balloons) {
        int n = balloons.size();
        List<Integer> arr = new ArrayList<>();
        arr.add(1);
        arr.addAll(balloons);
        arr.add(1);

        Map<Integer, Map<Integer, Integer>> dp = new HashMap<>();
        for (int i = 1; i <= n; i += 1) {
            Map<Integer, Integer> submap = new HashMap<>();
            for (int j = 1; j <= n; j += 1) {
                submap.put(j, null);
            }
            dp.put(i, submap);
        }

        return solve(arr, 1, n, dp);
    }

    public static void main(String[] args) {
        System.out.println(getMaximumCoins(Arrays.asList(3, 1, 5, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(7, 1, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(9, 1)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 2, 3, 4, 5)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5, 2, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(5, 10)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5)));
    }
}

class TabulationDP51 {
    /**
     * Time complexity is O(n^3) and space complexity is O(n^2).
     */
    public static Integer getMaximumCoins(List<Integer> balloons) {
        int n = balloons.size();
        List<Integer> arr = new ArrayList<>();
        arr.add(1);
        arr.addAll(balloons);
        arr.add(1);

        Map<Integer, Map<Integer, Integer>> dp = new HashMap<>();
        for (int i = 0; i <= n + 1; i += 1) {
            Map<Integer, Integer> submap = new HashMap<>();
            for (int j = 0; j <= n + 1; j += 1) {
                if (i > j) {
                    submap.put(j, 0);
                } else {
                    submap.put(j, Integer.MIN_VALUE);
                }
            }
            dp.put(i, submap);
        }

        for (int i = n; i >= 1; i -= 1) {
            for (int j = 1; j <= n; j += 1) {
                if (i > j) {
                    continue;
                }
                Integer maxCoins = Integer.MIN_VALUE;
                for (int index = i; index <= j; index += 1) {
                    Integer coins = arr.get(i - 1) * arr.get(index) * arr.get(j + 1) + dp.get(i).get(index - 1) + dp.get(index + 1).get(j);
                    maxCoins = Math.max(coins, maxCoins);
                }
                Map<Integer, Integer> mp = dp.get(i);
                mp.put(j, maxCoins);
                dp.put(i, mp);
            }
        }

        return dp.get(1).get(n);
    }

    public static void main(String[] args) {
        System.out.println(getMaximumCoins(Arrays.asList(3, 1, 5, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(7, 1, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(9, 1)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 2, 3, 4, 5)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5, 2, 8)));
        System.out.println(getMaximumCoins(Arrays.asList(5, 10)));
        System.out.println(getMaximumCoins(Arrays.asList(1, 5)));
    }
}